server/svix-server/src/v1/endpoints/attempt.rs

```diff
@@ -252,21 +220,256 @@ async fn list_attempts_by_endpoint(
+
+    if let Some(iterator) = iterator {
+        query = query.filter(messageattempt::Column::Id.lt(iterator));
+    }
+
+    query = list_attempts_by_endpoint_or_message_query_filters(query, query_params);
+
     Ok(Json(MessageAttemptOut::list_response(
         query.all(db).await?.into_iter().map(Into::into).collect(),
         limit as usize,
     )))
 }
 
+/// Flattens in a [`ListAttemptsByEndpointOrMsgQueryParameters`] and adds one extra query parameter
+#[derive(Debug, Deserialize, Validate)]
+pub struct ListAttemptsByMsgQueryParameters {
+    #[serde(flatten)]
+    #[validate]
+    common: ListAttemptsByEndpointOrMsgQueryParameters,
+
+    #[validate]
+    endpoint_id: Option<EndpointIdOrUid>,
+}
+
+/// Fetches a list of [`MessageAttemptOut`]s for a given message ID
+async fn list_attempts_by_msg(
+    Extension(ref db): Extension<DatabaseConnection>,
+    ValidatedQuery(mut pagination): ValidatedQuery<Pagination<MessageAttemptId>>,
+    ValidatedQuery(ListAttemptsByMsgQueryParameters {
+        common,
+        endpoint_id: endpoint_id_or_uid,
+    }): ValidatedQuery<ListAttemptsByMsgQueryParameters>,
+    Path((_app_id, msg_id)): Path<(ApplicationId, MessageId)>,
+    AuthenticatedApplication {
+        permissions: _,
+        app,
+    }: AuthenticatedApplication,
+) -> Result<Json<ListResponse<MessageAttemptOut>>> {
+    let limit = pagination.limit;
+    let iterator = pagination.iterator.take();
+
+    // Confirm message ID belongs to the given application
+    if message::Entity::secure_find_by_id(app.id.clone(), msg_id.clone())
+        .one(db)
+        .await?
+        .is_none()
+    {
+        return Err(Error::Http(HttpError::not_found(None, None)));
+    }
+
+    let mut query = messageattempt::Entity::secure_find_by_msg(msg_id)
+        .order_by_desc(messageattempt::Column::Id)
+        .limit(limit + 1);
+
+    if let Some(iterator) = iterator {
+        query = query.filter(messageattempt::Column::Id.lt(iterator));
+    }
+
+    if let Some(endpoint_id_or_uid) = endpoint_id_or_uid {
+        // Ensure the endpoint ID belongs to the given application
+        if let Some(endp) = endpoint::Entity::secure_find_by_id_or_uid(app.id, endpoint_id_or_uid)
+            .one(db)
+            .await?
+        {
+            query = query.filter(messageattempt::Column::EndpId.eq(endp.id));
+        } else {
+            return Err(Error::Http(HttpError::not_found(None, None)));
+        }
+    }
+
+    query = list_attempts_by_endpoint_or_message_query_filters(query, common);
+
+    Ok(Json(MessageAttemptOut::list_response(
+        query.all(db).await?.into_iter().map(Into::into).collect(),
+        limit as usize,
+    )))
+}
+
+/// A type combining information from [`messagedestination::Model`]s and [`endpoint::Model`]s to
+/// output information on attempted destinations
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, ModelOut)]
+#[serde(rename_all = "camelCase")]
+struct MessageEndpointOut {
```

Comment:
This one should use flatten and include `MessageOut` like you did for list attempted messages.